### PR TITLE
#167377530 Setup unit testing

### PR DESCRIPTION
**Pull request by ibkadeeko**
#### What does this PR do?
This pull request sets up unit testing using `mocha` `chai` and `chai-http` 

#### Description of Task to be completed
- install `chai-http`
- install `regenerator-runtime` to handle async tests
- create and configure `.nycrc` file
- change test script

#### How should this be manually tested?
N/A

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#167377530](https://pivotaltracker.com/story/show/167377530)

#### Screenshots (if appropriate)
Screenshot 2019-07-25 at 15 17 27

#### Questions:
N/A

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -24,38 +25,12 @@ app.get('/', (request, response) => {
...
-    });
-  });
-}
-
-app.use('*', (request, response) => {
+app.all('*', (request, response) => {
   response.status(404).send('Not Found');
 });
+
+app.use(errorHandler);
```
Comment (Merkll) [inline src/index.js:38]: Good job with the error handler. It Helps us tests better

File: tests/middleware.test.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { expect } from 'chai';
+import errorHandler from '../src/middlewares/errorHandler';
+
+describe('Error Handler', () => {
+  let req, res, next;
+  beforeEach(() => {
+    res = {
+      status(code) {
```
Comment (Merkll) [inline tests/middleware.test.js:14]: Nice Mocking job express req and res objects

**Review by Merkll — APPROVED** [review]
Good Job

File: tests/middleware.test.js (hunk 2 above)
Comment (Nennyfills) [reply tests/middleware.test.js:14]: Nice work right here!

**Review by Nennyfills — APPROVED** [review]
Great work

File: .env.example [hunk 3]
```diff
@@ -1,4 +1,2 @@
-DATABASE_URL=
```
Comment (iverenshaguy) [inline .env.example]: Please replace the environmental variables by making your .env file up to date

File: package.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
     "start": "node dist/index.js",
     "start:dev": "cross-env DEBUG=dev nodemon --exec babel-node src/index.js",
     "sync-env": "sync-dotenv",
-    "test": "mocha --require @babel/register tests/test.js --exit",
+    "test": "nyc mocha tests/test.js --exit",
```
Comment (iverenshaguy) [inline package.json]: Please remove the `nyc` from this test and use the `test:coverage` script for testing with coverage. Update the pre-commit hook to use `test:coverage` too

File: src/index.js [hunk 5]
```diff
@@ -24,38 +25,12 @@ app.get('/', (request, response) => {
...
-      errors: {
-        message: err.message,
-        error: {}
-      }
-    });
-  });
-}
-
-app.use('*', (request, response) => {
+app.all('*', (request, response) => {
```
Comment (iverenshaguy) [inline src/index.js:31]: Kindly explain the difference between `.all` and `.use`

File: src/middlewares/errorHandler.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import Debug from 'debug';
+import { config } from 'dotenv';
+
+config();
+const debug = Debug('dev');
+export default (err, request, response, next) => {
+  const isProduction = process.env.NODE_ENV === 'production';
+
+  let errorMessage = {};
+  if (!isProduction) {
```
Comment (iverenshaguy) [inline src/middlewares/errorHandler.js:15]: Please leave a new line between lines 9 and 10

File: src/middlewares/errorHandler.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    debug(err.stack);
+    errorMessage = err;
+  }
+  response.status(err.status || 500);
```
Comment (iverenshaguy) [inline src/middlewares/errorHandler.js]: Please leave a new line between lines 13 and 14

File: src/middlewares/errorHandler.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  response.json({
+    errors: {
+      message: err.message,
+      error: errorMessage
+    }
+  });
+  next();
```
Comment (iverenshaguy) [inline src/middlewares/errorHandler.js]: Is there a specific reason why we are still calling `next()` after sending the response to the user?

File: src/middlewares/errorHandler.js (hunk 7 above)
Comment (iverenshaguy) [reply src/middlewares/errorHandler.js]: Also kindly chain .status() and .json() and return the response. We want to avoid getting `headers already sent` errors.

```suggestion
  return response
    .status(err.status || 500)
    .json({
       errors: {
         message: err.message,
         error: errorMessage
       }
     });
```

File: tests/middleware.test.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    expect(res.status).to.be.eql(500);
+  });
+  it('should return specified error for production', async () => {
+    process.env.NODE_ENV = 'production';
+    const err = new Error('403 Error');
+    err.status = 403;
+    errorHandler(err, req, res, next);
+    expect(res.status).to.be.eql(403);
+  });
+  it('should return specified error for development', async () => {
```
Comment (iverenshaguy) [inline tests/middleware.test.js]: Is there a way to test that it returns the stack trace for development?

File: tests/middleware.test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      json() {}
+    };
+    req = {};
+    next = () => {};
+  });
+  it('should return 500 error status', async () => {
+    errorHandler(new Error('Test Error'), req, res, next);
+    expect(res.status).to.be.eql(500);
+  });
+  it('should return specified error for production', async () => {
```
Comment (iverenshaguy) [inline tests/middleware.test.js]:
```suggestion
it('should return specified error for production', async () => {
    process.env.NODE_ENV = 'production';
    const err = new Error('403 Error'); // won't be there when we define err out the test blocks
    err.status = 403;
 
    errorHandler(err, req, res, next);
 
    expect(res.status).to.be.eql(403);
  });
```
Please, try to do something like this for each block, let the `expect(...)` be after a new line so that it is easy to understand the flow of the tests instead of mushing everything up.

We can use this format for our test blocks;
```
[declarations and assignments] e.g. err.status = 403

[effects i.e. what is actually causing the effect] e.g.  errorHandler(err, req, res, next);

[results] e.g. expect(res.status).to.be.eql(403);
```

You can also define `err` outside the test blocks i.e.

```suggestion
describe('Error Handler', () => {
  let req, res, next;
  const err = new Error('Error');
  ...
```
and then change the error status in each test block. I also think it would be better to use `error` instead of `err`.

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
Really really good job, I should just approve it but for the sake of readability, kindly review my comments

File: src/index.js (hunk 5 above)
Comment (ibkadeeko) [reply src/index.js:31]: app.use() attaches to a middleware but app.all() is attached to the router.
We want the app.all to catch all routes that don't exist that are requested for.

File: src/middlewares/errorHandler.js (hunk 8 above)
Comment (ibkadeeko) [reply src/middlewares/errorHandler.js]: The way express recognises an error handler is when it has 4 parameters `err` `req` `res` `next`. Since the errorhandler is the last line in the hierarchy there really is no need to call it, but I have to declare `next` as one of the parameters so I called it so as for it to pass the lint rules.

File: tests/middleware.test.js (hunk 9 above)
Comment (ibkadeeko) [reply tests/middleware.test.js]: I'm still levelling up on sinon, so for now, I only tested the error handler just as a function not as a request.

File: .env.example (hunk 3 above)
Comment (ibkadeeko) [reply .env.example]: ok

File: package.json (hunk 4 above)
Comment (ibkadeeko) [reply package.json]: I used `nyc` to run the test so that the test script can be clean and I can do all the necessary configurations in the `.nycrc` file.

File: src/middlewares/errorHandler.js (hunk 8 above)
Comment (iverenshaguy) [reply src/middlewares/errorHandler.js]: This makes sense, you can add this logic to check if response headers have already been sent

```javascript
    if (res.headersSent) {
      return next(err);
    }
```

It uses next and handles an edge case. It's also inside the express documentation

File: tests/app.test.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    const response = await chai.request(app).get('/');
+    response.should.have.status(200);
+    response.body.should.be.a('object');
+  });
+});
+
+describe('Error Messages', () => {
+  it('should display 404 error on invalid route', async () => {
+    const res = await chai.request(app).get('/invalid/route');
+    res.should.have.status(404);
```
Comment (iverenshaguy) [inline tests/app.test.js]: We have to decide to use either should or expect. This will help us achieve consistency. Also, please leave a new line between 19 and 20.

File: tests/middleware.test.js (hunk 2 above)
Comment (iverenshaguy) [reply tests/middleware.test.js:14]: This is amazing work!

File: tests/middleware.test.js (hunk 9 above)
Comment (iverenshaguy) [reply tests/middleware.test.js]: Yeah, I understand but Sinon isn't exactly for requests. I mean, test that `res.json.errorMessage` is not an empty object in development and is an empty object in production i.e.

In development,
```javascript
expect(res.json.errorMessage).not.to.be.eql({});
```

In production,
```javascript
expect(res.json.errorMessage).to.be.eql({});
```

File: src/middlewares/errorHandler.js (hunk 8 above)
Comment (ibkadeeko) [reply src/middlewares/errorHandler.js]:
```javascript
    if (res.headersSent) {
      return next(err);
    }
```
This code is unreachable once i return the response

**Review by iverenshaguy — APPROVED** [review]
